### PR TITLE
fix appsi gurobi available()

### PR DESCRIPTION
**Pull request by michaelbynum**
## Summary/Motivation:
This PR updates the `available()` method in the appsi-Gurobi interface so that license availability can be checked multiple times. 

### Legal Acknowledgement

By contributing to this software project, I have read the [contribution guide](https://pyomo.readthedocs.io/en/stable/contribution_guide.html) and agree to the following terms and conditions for my contribution:

1. I agree my contributions are submitted under the BSD license.
2. I represent I am authorized to make the contributions and grant the license. If my employer has rights to intellectual property that includes these contributions, I represent that I have received permission to make contributions and grant the required license on behalf of that employer.

### CODE REVIEW

**Comment by michaelbynum** [comment]
This has some undesired side effects...

**Comment by jsiirola** [comment]
Why does this remove the code to distinguish between demo and full licenses?  More generally, I think I would prefer to leave the default interface to rely on caching and instead add either an option or a new method to forget the previously cached result?

**Comment by michaelbynum** [comment]
I don't think caching is a good idea when the result can change.

**Comment by michaelbynum** [comment]
I guess we could cache demo vs full, but recheck availability each call. Demo vs full should not change.

File: pyomo/contrib/appsi/solvers/gurobi.py [hunk 1]
```diff
@@ -257,42 +257,46 @@ def __init__(self, only_child_vars=True):
...
             with capture_output(capture_fd=True):
                 m = gurobipy.Model()
-                m.dispose()
-        except ImportError:
-            # Triggered if this is the first time the deferred import of
-            # gurobipy is resolved. _import_gurobipy will have already
-            # set _available appropriately.
-            return
+            if self._solver_model is None:
+                self._solver_model = m
```
Comment (jsiirola) [inline pyomo/contrib/appsi/solvers/gurobi.py:275]: This looks like the solver interface will create (and perpetually hold on to) a Gurobi model -- I suspect to grab the license and keep others from "stealing" it from us.  Can you document somewhere (preferably obvious for users) the strategy that we are using for managing Gurobi licenses (and in particular what a user should do to release their license)?

Comment (michaelbynum) [reply pyomo/contrib/appsi/solvers/gurobi.py:275]: Done!

**Review by mrmundt — APPROVED** [review]

**Review by jsiirola — APPROVED** [review]

**Pull request merged**